File: miniapp/backend/routers/watchlist.py
from typing import List, Optional

from fastapi import APIRouter, Depends
from pydantic import BaseModel
from sqlalchemy.orm import Session

from core.responses import success_response
from crud.watchlist import (
    batch_update_watchlist_active,
    create_watchlist,
    delete_watchlist,
    list_watchlist,
    update_watchlist,
)
from deps import get_current_user, get_db
from models import User
from schemas.watchlist import Watchlist as WatchlistSchema
from schemas.watchlist import WatchlistCreate, WatchlistUpdate

router = APIRouter(prefix="/api/watchlist", tags=["watchlist"])


class WatchlistBatchUpdateRequest(BaseModel):
    ids: List[int]
    is_active: bool


@router.get("")
def get_watchlist(
    active: Optional[bool] = None,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    entries = list_watchlist(db, current_user.id, is_active=active)
    return success_response([WatchlistSchema.model_validate(item) for item in entries])


@router.post("")
def create_watchlist_item(
    payload: WatchlistCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    entry = create_watchlist(
        db,
        current_user.id,
        payload.protocol_name,
        payload.condition_type,
        payload.threshold,
        is_active=payload.is_active,
        alert_message=payload.alert_message,
    )
    return success_response(WatchlistSchema.model_validate(entry))


@router.patch("/batch")
def batch_update_watchlist(
    payload: WatchlistBatchUpdateRequest,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    updated, missing = batch_update_watchlist_active(
        db, current_user.id, payload.ids, payload.is_active
    )
    return success_response({"updated": updated, "missing": missing})


@router.patch("/{watchlist_id}")
def update_watchlist_item(
    watchlist_id: int,
    payload: WatchlistUpdate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    entry = update_watchlist(
        db,
        current_user.id,
        watchlist_id,
        condition_type=payload.condition_type,
        threshold=payload.threshold,
        is_active=payload.is_active,
        alert_message=payload.alert_message,
    )
    return success_response(WatchlistSchema.model_validate(entry))


@router.delete("/{watchlist_id}")
def delete_watchlist_item(
    watchlist_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    delete_watchlist(db, current_user.id, watchlist_id)
    return success_response({"deleted": True, "id": watchlist_id})

File: miniapp/frontend/src/services/watchlist.ts
import { apiClient } from './api'

export type WatchlistItem = {
  id: number
  protocolName: string
  conditionType: string
  threshold: number
  isActive: boolean
  alertMessage?: string | null
  lastNotified?: string | null
  createdAt: string
  updatedAt: string
}

type RawWatchlist = {
  id: number
  protocol_name: string
  condition_type: string
  threshold: number
  is_active: boolean
  alert_message?: string | null
  last_notified?: string | null
  created_at: string
  updated_at: string
}

const mapWatchlist = (item: RawWatchlist): WatchlistItem => ({
  id: item.id,
  protocolName: item.protocol_name,
  conditionType: item.condition_type,
  threshold: item.threshold,
  isActive: item.is_active,
  alertMessage: item.alert_message ?? null,
  lastNotified: item.last_notified ?? null,
  createdAt: item.created_at,
  updatedAt: item.updated_at,
})

export const getWatchlist = async (
  active?: boolean,
): Promise<WatchlistItem[]> => {
  const response = await apiClient.get<{ data: RawWatchlist[] }>(
    '/api/watchlist',
    { params: active !== undefined ? { active } : undefined },
  )
  return (response.data.data ?? []).map(mapWatchlist)
}

export const createWatchlist = async (payload: {
  protocolName: string
  conditionType: string
  threshold: number
  isActive?: boolean
  alertMessage?: string
}): Promise<WatchlistItem> => {
  const response = await apiClient.post<{ data: RawWatchlist }>('/api/watchlist', {
    protocol_name: payload.protocolName,
    condition_type: payload.conditionType,
    threshold: payload.threshold,
    is_active: payload.isActive ?? true,
    alert_message: payload.alertMessage,
  })
  return mapWatchlist(response.data.data)
}

export const updateWatchlist = async (
  id: number,
  payload: {
    conditionType?: string
    threshold?: number
    isActive?: boolean
    alertMessage?: string | null
  },
): Promise<WatchlistItem> => {
  const response = await apiClient.patch<{ data: RawWatchlist }>(
    `/api/watchlist/${id}`,
    {
      condition_type: payload.conditionType,
      threshold: payload.threshold,
      is_active: payload.isActive,
      alert_message: payload.alertMessage,
    },
  )
  return mapWatchlist(response.data.data)
}

export const batchUpdateWatchlist = async (ids: number[], isActive: boolean) => {
  await apiClient.patch('/api/watchlist/batch', { ids, is_active: isActive })
  return true
}

export const deleteWatchlist = async (id: number) => {
  await apiClient.delete(`/api/watchlist/${id}`)
  return true
}
